src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -561,19 +557,21 @@ grpc_chttp2_transport::grpc_chttp2_transport(
 
 static void destroy_transport_locked(void* tp, grpc_error* error) {
   grpc_chttp2_transport* t = static_cast<grpc_chttp2_transport*>(tp);
+  gpr_mu_lock(&t->mu);
   t->destroying = 1;
   close_transport_locked(
       t, grpc_error_set_int(
              GRPC_ERROR_CREATE_FROM_STATIC_STRING("Transport destroyed"),
              GRPC_ERROR_INT_OCCURRED_DURING_WRITE, t->write_state));
+  gpr_mu_unlock(&t->mu);
   // Must be the last line.
   GRPC_CHTTP2_UNREF_TRANSPORT(t, "destroy");
 }
 
 static void destroy_transport(grpc_transport* gt) {
   grpc_chttp2_transport* t = reinterpret_cast<grpc_chttp2_transport*>(gt);
   GRPC_CLOSURE_SCHED(GRPC_CLOSURE_CREATE(destroy_transport_locked, t,
```

Comment:
Is this closure still necessary?  It seems like if we no longer need to bounce into the combiner, we should be able to avoid this level of indirection.  I think we can just move the code from `destroy_transport_locked()` directly into this function.

It looks like there are a number of cases like this throughout the code.

Reply:
I agree. I was thinking that I will do this iteratively, so that this PR is small, but I can add these changes

Reply:
@markdroth , I've removed this and some obvious cases. There are a few remaining, particularly around writes (write_action, write_action_begin). Changing these would have an impact on timing and would probably have an impact on performance. I would prefer to do these in a separate PR so that we can measure it.